### PR TITLE
Blas/team policy dot test

### PR DESCRIPTION
**Pull request by ajpowelsnl**
The proposed code represents a TeamPolicy-based dot performance test (https://github.com/kokkos/kokkos-kernels/wiki/BLAS-1%3A%3Ateam-dot).  In the near future, this test will also be implemented with RAJAPerf Suite infrastructure for nightly builds, performance testing and reporting.  

(SAMPLE RESULTS WITH OPENMP IMPLEMENTATION; OTHER CASES SHOWN TO DEMO EXPECTED ERRORS) 

 ./KokkosBlas_team_dot_perf_test 
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
ERROR: Serial device requested, but not available; here, implementation of dot is explicitly parallel.
[ajpowel@kokkos-dev-2 blas]$ 
[ajpowel@kokkos-dev-2 blas]$ 
[ajpowel@kokkos-dev-2 blas]$ ./KokkosBlas_team_dot_perf_test --openmp 20
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Running BLAS Level 1 Kokkos Teams-based implementation DOT performance experiment (OpenMP)
Each test input vector has a length of 100000
Avg DOT time: 0.000172 s.
Avg DOT FLOP/s: 1.164e+09
[ajpowel@kokkos-dev-2 blas]$ ./KokkosBlas_team_dot_perf_test --threads 20
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
ERROR:  PThreads requested, but not available.
[ajpowel@kokkos-dev-2 blas]$ 
[ajpowel@kokkos-dev-2 blas]$ 
[ajpowel@kokkos-dev-2 blas]$ ./KokkosBlas_team_dot_perf_test --cuda 0
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
ERROR: CUDA requested, but not available.

### CODE REVIEW

File: perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+    if (useOMP) {
+#if defined(KOKKOS_ENABLE_OPENMP)
+      if (params.layoutLeft)
+        run<Kokkos::OpenMP, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::OpenMP, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: OpenMP requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp:233]: Minor: indentation is misaligned L233-234

File: perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    if (useCUDA) {
+#if defined(KOKKOS_ENABLE_CUDA)
+      if (params.layoutLeft)
+        run<Kokkos::Cuda, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::Cuda, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: CUDA requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp:245]: Minor: indentation is misaligned L245-246

File: perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+#endif
+    }
+    if (useSerial) {
+#if defined(KOKKOS_ENABLE_SERIAL)
+      if (params.layoutLeft)
+        run<Kokkos::Serial, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::Serial, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: Serial device requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_dot_perf_test.cpp:256]: Minor: indentation is misaligned L256-267

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+TeamVectorRange 	Used inside of a TeamPolicy kernel to perform nested parallel loops split over threads of a team and their vector lanes.
+ThreadVectorRange 	Used inside of a TeamPolicy kernel to perform nested parallel loops with vector lanes of a thread.
+*/
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+template <class ExecSpace, class Layout>
+void run(int m, int repeat) {
+  // Declare type aliases
+    using Scalar   = double;
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:174]: Minor: indentation is misaligned L174-180

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+  if (useThreads)
+  {
+#if defined(KOKKOS_ENABLE_THREADS)
+    if (params.use_threads)
+      run<Kokkos::Threads, Kokkos::LayoutLeft>(params.m, params.repeat);
+    else
+      run<Kokkos::Threads, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+    std::cout << "ERROR:  PThreads requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:259]: Minor: indentation is misaligned L259-260,262

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+    if (useOMP) {
+#if defined(KOKKOS_ENABLE_OPENMP)
+      if (params.layoutLeft)
+        run<Kokkos::OpenMP, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::OpenMP, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: OpenMP requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:271]: Minor: indentation is misaligned L271-2

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    if (useCUDA) {
+#if defined(KOKKOS_ENABLE_CUDA)
+      if (params.layoutLeft)
+        run<Kokkos::Cuda, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::Cuda, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: CUDA requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:283]: Minor: indentation is misaligned L283-4

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+#endif
+    }
+    if (useSerial) {
+#if defined(KOKKOS_ENABLE_SERIAL)
+      if (params.layoutLeft)
+        run<Kokkos::Serial, Kokkos::LayoutLeft>(params.m, params.repeat);
+      else
+        run<Kokkos::Serial, Kokkos::LayoutRight>(params.m, params.repeat);
+#else
+  std::cout << "ERROR: Serial device requested, but not available; here, implementation of dot is explicitly parallel.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:294]: Minor: indentation is misaligned L294-5

File: perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+                  KOKKOS_LAMBDA(const member_type& team){
+                       });
+
+  // To guarantee a kernel has finished, a developer should call the fence of the execution space on which the kernel is being executed. 
+  // Otherwise, it depends on the execution space where the loop executes, and whether this execution space implements a barrier.
+
+  Kokkos::fence();
+  Kokkos::Timer timer;
+
+  // Live test of dot:
```
Comment (ndellingwood) [inline perf_test/blas/blas/KokkosBlas_team_dot_perf_test.cpp:212]: Missing the repeat loop around the live test

File: perf_test/blas/blas2/KokkosBlas2_gemv_perf_test.cpp [hunk 10]
```diff
@@ -116,52 +116,115 @@ int parse_inputs (Params& params, int argc, char **argv){
...
+  // Logic for runtime with PThreads 
+    if (useThreads)
+  {
+#if defined(KOKKOS_ENABLE_THREADS)
+    if (params.use_threads)
+      run<Kokkos::Threads, Kokkos::LayoutLeft>(params.m, params.n, params.repeat);
+    else
+      run<Kokkos::Threads, Kokkos::LayoutRight>(params.m, params.n, params.repeat);
+#else
+    std::cout << "ERROR:  PThreads requested, but not available.\n";
```
Comment (ndellingwood) [inline perf_test/blas/blas2/KokkosBlas2_gemv_perf_test.cpp:222]: Minor: indentation is misaligned L222-5

**Review by ndellingwood — COMMENTED** [review]
@ajpowelsnl nice PR :) only missing element is a repeat loop wrapping the live team dot test; everything else is just minor indentation alignment notes. Thanks!

**Comment by kokkos-devops-admin** [comment]
Status Flag 'Pre-Test Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           THE LAST COMMIT TO THIS PULL REQUEST HAS BEEN REVIEWED, BUT NOT ACCEPTED OR REQUIRES CHANGES

**Pull request closed without merging**